crates/wasmparser/src/readers/core/types/matches.rs

```diff
@@ -0,0 +1,380 @@
+                            WithRecGroup::map(a, |_| *pa),
+                            offset,
+                        )?)
+                })?;
+        if !params_match {
+            return Ok(false);
+        }
+
+        let results_match =
+            a.results()
+                .iter()
+                .zip(b.results())
+                .try_fold(true, |matches, (ra, rb)| {
+                    // Results are covariant.
+                    Ok(matches
+                        && types.matches(
+                            WithRecGroup::map(a, |_| *ra),
+                            WithRecGroup::map(b, |_| *rb),
+                            offset,
+                        )?)
+                })?;
+        Ok(results_match)
+    }
+}
+
+impl Matches for WithRecGroup<ArrayType> {
+    fn matches(types: &TypeList, a: Self, b: Self, offset: usize) -> Result<bool> {
+        types.matches(
+            WithRecGroup::map(a, |a| a.0),
+            WithRecGroup::map(b, |b| b.0),
+            offset,
+        )
+    }
+}
+
+impl<'a> Matches for WithRecGroup<&'a StructType> {
+    fn matches(types: &TypeList, a: Self, b: Self, offset: usize) -> Result<bool> {
+        // Note: Struct types support width and depth subtyping.
+        Ok(a.fields.len() >= b.fields.len()
+            && a.fields
+                .iter()
+                .zip(b.fields.iter())
+                .try_fold(true, |matches, (fa, fb)| {
+                    Ok(matches
+                        && types.matches(
+                            WithRecGroup::map(a, |_| *fa),
+                            WithRecGroup::map(b, |_| *fb),
+                            offset,
+                        )?)
+                })?)
+    }
+}
+
+impl Matches for WithRecGroup<FieldType> {
+    fn matches(types: &TypeList, a: Self, b: Self, offset: usize) -> Result<bool> {
+        Ok((b.mutable || !a.mutable)
+            && types.matches(
+                WithRecGroup::map(a, |a| a.element_type),
+                WithRecGroup::map(b, |b| b.element_type),
+                offset,
+            )?)
+    }
+}
+
+impl Matches for WithRecGroup<StorageType> {
+    fn matches(types: &TypeList, a: Self, b: Self, offset: usize) -> Result<bool> {
+        use StorageType as ST;
+        match (*a, *b) {
+            (ST::I8, ST::I8) | (ST::I16, ST::I16) => Ok(true),
+            (ST::Val(va), ST::Val(vb)) => types.matches(
+                WithRecGroup::map(a, |_| va),
+                WithRecGroup::map(b, |_| vb),
+                offset,
+            ),
+            _ => Ok(false),
+        }
+    }
+}
+
+impl Matches for WithRecGroup<ValType> {
+    fn matches(types: &TypeList, a: Self, b: Self, offset: usize) -> Result<bool> {
+        match (*a, *b) {
+            (ValType::Ref(ra), ValType::Ref(rb)) => types.matches(
+                WithRecGroup::map(a, |_| ra),
+                WithRecGroup::map(b, |_| rb),
+                offset,
+            ),
+            (a, b) => Ok(a == b),
```

Comment:
Could this expand `a` to `a @ (ValType::I32 | ValType::I64 ...)` to avoid the catch-all?